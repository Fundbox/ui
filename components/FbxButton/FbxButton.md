# FbxButton

## Basic usage
- Pass any HTML attribute or native DOM event to `fbx-button`
- When `:loading` is `true`, a spinning icon will replace the button text, and the user will not be able to click the button again until loading has finished and is `false`
- Button text is passed as a `<slot>`

```html
<fbx-button
    name="my-button"
    id="my-id"
    data-qa="my-button-qa"
    tabindex="3"
    @click="onClick"
    :loading="isSubmittingForm"
>
  My Button Text
</fbx-button>
```

## Disabled button

- Use `disable` like you would on a normal HTML button. Note: `disabled` here is not a Vue prop, but a regular HTML attribute

```html
<fbx-button
  @click="onClick"
  :disabled="shouldButtonBeDisabled"
>
  My Button Text
</fbx-button>
```

## Inverse button

- Reverse the background color and font color by passing `true` for the `:inverse` prop
- Do not use this prop if you just want a regular `<a>` element

```html
<fbx-button
  @click="onClick"
  :inverse="true"
>
  My Button Text
</fbx-button>
```
