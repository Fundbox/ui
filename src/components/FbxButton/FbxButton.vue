<template>
  <button
    tabindex="0"
    type="button"
    class="fbx-cta-button"
    :class="{'fbx-button-inverse': inverse, 'loading': loading }"
    :disabled="loading"
    v-on="listeners"
  >
    <span v-if="loading" class="fbx-button-loading"></span>
    <slot v-else></slot>
  </button>
</template>

<script>
export default {
  name: 'FbxButton',
  props: {
    loading: {
      type: Boolean,
      default: false
    },
    inverse: {
      type: Boolean,
      default: false
    }
  },
  computed: {
    listeners() {
      return {
        ...this.$listeners
      }
    }
  }
}
</script>

<style lang="scss" scoped>
@import "../../styles/utils/mixins.scss";
@import "../../styles/utils/variables.scss";
@import "../../styles/utils/color-palette.scss";

.fbx-cta-button {
  position: relative;
  display: block;
  min-height: 30px;
  min-width: 120px;
  max-width: 100%;
  height: $input-height;
  padding: 10px 20px;
  overflow: hidden;
  transition: background $super-fast-ease;
  font-size: 14px;
  font-weight: 300;
  // Prevent font mixin from making the button text uncentered
  line-height: normal !important;
  color: $white;
  text-align: center;
  outline: none;
  border: 0;
  // Chrome 62 adds 4px border-radius by default to match MacOS native styles
  border-radius: 0;
  cursor: pointer;
  user-select: none;
  white-space: nowrap;
  box-sizing: border-box;
  background: $dark-green;
  @include viewport("md") {
    &:hover:not(:disabled) {
      background: $medium-green;
    }
    &:active:not(:disabled) {
      background: $dark-green;
    }
  }

  &:disabled:not(.loading) {
    color: $extra-dark-gray;
    border: 1px solid $extra-dark-gray;
    background: $white;
    cursor: default;
  }

  &:focus {
    outline: none;
  }

  &.fbx-button-inverse {
    background: transparent;
    color: $dark-green;
    @include viewport("md") {
      &:hover {
        background: $dark-green;
        color: $white;
      }

      &:active {
        background: $light-green;
        color: $white;
      }
    }
  }
}

.fbx-button-loading {
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  display: block;
  content: "";
  position: absolute;
  background-image: url("../../assets/btn-ajax-loading.gif");
  background-position: center;
  background-repeat: no-repeat;
  background-color: transparent;
  background-size: 40px;
}
</style>
