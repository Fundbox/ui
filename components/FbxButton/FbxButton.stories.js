import { storiesOf } from '@storybook/vue';
import { action } from '@storybook/addon-actions';
import { withInfo } from 'storybook-addon-vue-info';
import { text, boolean } from '@storybook/addon-knobs';

import FbxButton from './FbxButton.vue';
import summary from './FbxButton.md';

const stories = storiesOf('FbxButton', module);

stories.add('Default', withInfo({ summary })(() => ({
  components: { FbxButton },
  template: `<fbx-button
                @click="click"
                :loading="${boolean('Loading', false)}"
                ${boolean('Disabled', false) ? 'disabled' : ''}
              >${text('Text', 'Click me')}</fbx-button>`,
  methods: { click: action('clicked') },
})));

stories.add('with HTML attributes', () => ({
  components: { FbxButton },
  template: `<fbx-button
              name="fbx-button"
              id="my-id"
              data-qa="button-qa"
              tabindex="3"
            >Click me</fbx-button>`,
  methods: { click: action('clicked') },
}));
