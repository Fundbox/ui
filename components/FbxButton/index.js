import Vue from "vue";
import FbxButton from "./FbxButton.vue";
const vComponent = Vue.component(FbxButton.name, FbxButton);
export default vComponent;
