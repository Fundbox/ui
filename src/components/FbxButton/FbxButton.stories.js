import { storiesOf } from '@storybook/vue'
import { action } from '@storybook/addon-actions'
import { withInfo } from 'storybook-addon-vue-info'
import { text, boolean } from '@storybook/addon-knobs'

import FbxButton from './FbxButton.vue'
import summary from './FbxButton.md'

const stories = storiesOf('Components/Button', module)

stories.add('default', withInfo({ summary })(() => ({
  components: { FbxButton },
  template: `<fbx-button
                name="fbx-button"
                id="my-id"
                data-qa="button-qa"
                tabindex="3"
                @click="click"
                :loading="${boolean('Loading', false)}"
                :inverse="${boolean('Inverse', false)}"
                :disabled="${boolean('Disabled', false)}"
            >
              ${text('Text', 'Click me')}
            </fbx-button>`,
  methods: { click: action('clicked') },
})))
