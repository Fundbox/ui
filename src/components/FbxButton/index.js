import FbxButton from './FbxButton.vue';
import { registerComponents, vueUse } from '../../utils/plugins';

const components = {
  FbxButton
};

const VuePlugin = {
  install(Vue) {
    registerComponents(Vue, components);
  }
};

vueUse(VuePlugin);

export default VuePlugin;
