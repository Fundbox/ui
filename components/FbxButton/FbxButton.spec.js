import { shallowMount } from "@vue/test-utils"
import FbxButton from "./FbxButton.vue"

const buttonText = "Submit"
const defaultSlot = { default: buttonText };

describe("FbxButton", () => {
  describe("snapshots", () => {
    it("renders the default correctly", () => {
      const wrapper = shallowMount(FbxButton, {
        slots: defaultSlot,
      })

      expect(wrapper.html()).toMatchSnapshot()
    })

    it("renders loading correctly", () => {
      const wrapper = shallowMount(FbxButton, {
        propsData: { loading: true },
        slots: defaultSlot,
      })

      expect(wrapper.html()).toMatchSnapshot()
    })

    it("renders inverse correctly", () => {
      const wrapper = shallowMount(FbxButton, {
        propsData: { inverse: true },
        slots: defaultSlot,
      })

      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe("@events", () => {
    it("listens to any event listener you pass", () => {
      const mockClick = jest.fn()
      const mockHover = jest.fn()
      const wrapper = shallowMount(FbxButton, {
        listeners: {
          click: mockClick,
          hover: mockHover,
        }
      })

      wrapper.trigger("click")
      wrapper.trigger("hover")

      expect(mockClick).toHaveBeenCalledTimes(1)
      expect(mockHover).toHaveBeenCalledTimes(1)
    })
  })

  describe("disabled", () => {
    it("prohibits clicking again while loading", () => {
      const mockClick = jest.fn()
      const wrapper = shallowMount(FbxButton, {
        listeners: {
          click: mockClick,
        },
        slots: defaultSlot,
      })

      wrapper.trigger("click")
      wrapper.setProps({ loading: true })
      expect(mockClick).toHaveBeenCalledTimes(1)

      wrapper.trigger("click")
      expect(mockClick).toHaveBeenCalledTimes(1)

      wrapper.setProps({ loading: false })
      wrapper.trigger("click")
      expect(mockClick).toHaveBeenCalledTimes(2)
    })
  })
})
