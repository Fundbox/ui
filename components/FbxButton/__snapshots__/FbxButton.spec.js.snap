// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FbxButton snapshots renders inverse correctly 1`] = `<button tabindex="0" type="button" class="fbx-cta-button fbx-button-inverse">Submit</button>`;

exports[`FbxButton snapshots renders loading correctly 1`] = `<button tabindex="0" type="button" disabled="disabled" class="fbx-cta-button loading"><span class="fbx-button-loading"></span></button>`;

exports[`FbxButton snapshots renders the default correctly 1`] = `<button tabindex="0" type="button" class="fbx-cta-button">Submit</button>`;
